connectors/cache/fallback.go

```diff
@@ -0,0 +1,306 @@
+			if err != nil {
+				return err
+			}
+			newValues := map[string]dosa.FieldValue{
+				key:   cacheKey,
+				value: cacheValue,
+			}
+			return c.fallback.Upsert(ctx, adaptedEi, newValues)
+		}
+		_ = c.cacheWrite(w)
+
+		return sourceRows, sourceToken, sourceErr
+	}
+	value, err := c.getValueFromFallback(ctx, adaptedEi, cacheKey)
+	c.logFallback("RANGE", err)
+	if err != nil {
+		return sourceRows, sourceToken, sourceErr
+	}
+	unpack := rangeResults{}
+	err = c.encoder.Decode(value, &unpack)
+	if err != nil {
+		return sourceRows, sourceToken, sourceErr
+	}
+	return unpack.Rows, unpack.TokenNext, err
+}
+
+// Scan returns scan result from origin.
+func (c *Connector) Scan(ctx context.Context, ei *dosa.EntityInfo, minimumFields []string, token string, limit int) ([]map[string]dosa.FieldValue, string, error) {
+	// Scan will just call range with no conditions
+	return c.Range(ctx, ei, nil, minimumFields, token, limit)
+}
+
+// Remove deletes an entry
+func (c *Connector) Remove(ctx context.Context, ei *dosa.EntityInfo, keys map[string]dosa.FieldValue) error {
+	w := func() error {
+		cacheKey := createCacheKey(ei, keys, c.encoder)
+		adaptedEi := adaptToKeyValue(ei)
+		return c.fallback.Remove(ctx, adaptedEi, map[string]dosa.FieldValue{key: cacheKey})
+	}
+
+	if c.isCacheable(ei) {
+		_ = c.cacheWrite(w)
+	}
+
+	return c.origin.Remove(ctx, ei, keys)
+}
+
+func (c *Connector) getValueFromFallback(ctx context.Context, ei *dosa.EntityInfo, keyValue []byte) ([]byte, error) {
+	response, err := c.fallback.Read(ctx, ei, map[string]dosa.FieldValue{key: keyValue}, dosa.All())
+	if err != nil {
+		return nil, err
+	}
+
+	// unpack the value
+	cacheValue, ok := response[value].([]byte)
+	if !ok {
+		return nil, errors.New("No value in cache for key")
+	}
+	return cacheValue, nil
+}
+
+func (c *Connector) logFallback(method string, err error) {
+	if c.stats != nil {
+		s := c.stats.SubScope("fallback").Tagged(map[string]string{"method": method})
+		if err != nil {
+			s = s.SubScope("failure")
+		} else {
+			s = s.SubScope("success")
+		}
+
+		s.Counter("use_fallback").Inc(1)
+	}
+}
+
+func (c *Connector) setSynchronousMode(sync bool) {
+	c.synchronous = sync
+}
+
+func (c *Connector) cacheWrite(w func() error) error {
+	if c.synchronous {
+		return w()
+	}
+	go func() { _ = w() }()
+	return nil
+}
+
+func (c *Connector) isCacheable(ei *dosa.EntityInfo) bool {
+	return c.cacheableEntities[ei.Def.Name]
+}
+
+// returns a set of entity names that should go through the fallback cache
+// All other entities will just return results from origin
+func createCachedEntitiesSet(entities []dosa.DomainObject) map[string]bool {
+	set := map[string]bool{}
+	for _, e := range entities {
+		if e == nil {
+			continue
+		}
+		t, err := dosa.TableFromInstance(e)
+		if err != nil {
+			continue
+		}
+		set[t.EntityDefinition.Name] = true
+	}
+	return set
+}
+
+func adaptToKeyValue(ei *dosa.EntityInfo) *dosa.EntityInfo {
```

Comment:
suggest adding a comment here (or changing the name of the method) to describe that a new entity info is being derived from the original to support the structure of a caching connector